x/finality/keeper/msg_server.go

```diff
@@ -206,6 +214,7 @@ func (ms msgServer) CommitPubRandList(goCtx context.Context, req *types.MsgCommi
 		StartHeight: req.StartHeight,
 		NumPubRand:  req.NumPubRand,
 		Commitment:  req.Commitment,
+		EpochNum:    ms.BTCStakingKeeper.GetEpoch(ctx),
```

Comment:
So this is breaking change as described in https://babylonlabs.atlassian.net/wiki/spaces/BABYLON/pages/10945276/Babylon+Chain+upgrading+guidelines+Ready+for+review#Changing-DB-schema-of-the-existing-module

So in theory we should now do all the steps described here (bump consensus version, write migration etc)

But the thing is, we already know that database of the chain after TGE version will be almost empty (i.e only few transfers and module paramters), so maybe we can get away without doing this whole dance.

Could you do small reality check ie:
1. Create binary from TGE version https://github.com/babylonlabs-io/babylon/releases/tag/v0.9.0
2. Create binary from version in this pr
3. Sync few blocks on TGE version 
4. then try to do upgrade to the pr version by manual update - https://docs.cosmos.network/main/build/tooling/cosmovisor#adding-upgrade-binary
5. try to commit some randomness 

If this works this will prove, we do not need to write migrations/bump consensus versions for breaking changes in modules which db will be empty.

Reply:
Why do we need to bump consensus version for this? This is to dev branch and the version is not necessarily deployed on any devnet/testnet/mainnet. So maybe we only need to increment the consensus version if 1) we have a new tagged version, and 2) this new tagged version makes breaking changes to consensus

Also, can such a reality check be wrapped into a CI? Tbh I don't think I can do this properly in a manual way :sweat:

Reply:
>Why do we need to bump consensus version for this?

Because this is breaking change to the module schema and if this change is introduced it will break consensus. Remember, at this point we already 'released' mainnet (v0.9.0) and what we are working on is actually upgrade to this mainnet version.

>Also, can such a reality check be wrapped into a CI? Tbh I don't think I can do this properly in a manual way 

Ideally it would be. I imagine that:
- in this pr @gitferry will do the check manually
- after that all the steps will be written on confluence
- then we can dockerize this steps and automate them

Reply:
> Remember, at this point we already 'released' mainnet (v0.9.0) and what we are working on is actually upgrade to this mainnet version.

But this version is not going to be deployed as a new network that inherits the ledger of the "mainnet". So from user's pov this "mainnet" is not upgraded upon this PR is merged

Reply:
> For this specific version we probably don't need to bump the consensus version because when we run v0.9.0, the db related to btc staking would be empty. I need to do the reality checks to verify.

I was saying from v0.9.0 to phase-2 mainnet, the consensus version should only increment *once*. This is because none of the versions in between will go live. 

That is, even if this PR has breaking changes to the DB, if the consensus version is 1 bigger than the value in v0.9.0, then this PR should not increment the consensus version. As this PR is somehow "overrides" or "replaces" the version prior to this PR.

> I don't get it. Wouldn't the current version we are working on be the upgrade of v0.9.0?

Yes, but this does not mean every state-breaking PR needs to increment the consensus version. Rather, before the next public network (i.e., phase-2 mainnet), we keep the same consensus version which is 1 bigger than that in v0.9.0, and every PR overwrites the upgrade handler of the prior PR.

Maybe I'm missing some important context?

Reply:
>But this version is not going to be deployed as a new network that inherits the ledger of the "mainnet". So from user's pov this "mainnet" is not upgraded upon this PR is merged

I also do not understand.

Ultimately, due to this pr ( and probably bunch of others) the next version we release, lets call it `v0.10.0`, will be breaking when compared to `v0.9.0` this means few things:
1. Everybody wanting to sync whole chain from block 1 of the mainnet will need to have 2 binaries: `v0.9.0` binary and `v0.10.0` binary and cosmovisor to to the proper swap of binaries at proper upgrade block
2. Now we have two ways of doing devnets for testing:
- we can deploy devnet of only version v0.10.0 and do not have any governance proposal, then there is not binary switch and we are only testing latest version
- we can deploy devnet on which we produce few blocks on v0.9.0 and then then do the governance  proposal to switch to v0.10.0, this way we test switch to and latest version
3. All this makes it necessary to properly upgrade module consensus versions
4. One observation we can make is that, when working on this next version v0.10.0 we only need to update consensus version only once with first breaking change, as we only do one upgrade from v0.9.0 to v0.10.0

Reply:
>Yes, but this does not mean every state-breaking PR needs to increment the consensus version

I think we agree on this. Only first state breaking pr will need to:
- bump consensus version
- create migration function
- register migration function

Later state breaking pr-s, will need to only modify the migration function to adjust it for the made changes

Reply:
@SebastianElvis @KonradStaniec Thanks for the discussions. Makes a lot of sense! Let's wrap up the above to the upgrade guidelines on confluence as it is slightly out of scope of this pr imo.

Reply:
OK, I think now I get it. So

- Consensus version of a module is incremented at most once between every two consecutive major versions. 
- This increment is needed only when the new major version introduces breaking changes to the given module.
- The increment happens at the 1st PR that introduces the 1st breaking change between the two consecutive major versions.

Is that correct? If yes then this sounds good to me. 

Another thing is that Cosmos SDK uses consensus version to [identify](https://github.com/cosmos/cosmos-sdk/blob/main/docs/learn/advanced/15-upgrade.md#overwriting-genesis-functions) the state migration function. This basically means that upon incrementing a consensus version, one needs to define a new state migration function for migrating the prior version to the new version. Will we follow this practice or we will only maintain 1 big state migration function for each module?

Reply:
>Is that correct? If yes then this sounds good to me.

Yup, I will clarify it in guidelines so that we are on the same page

>  Will we follow this practice or we will only maintain 1 big state migration function for each module?

I would follow this practice tbh. This creates nice history of upgrades between versions and makes the code clearer imo.

So after phase-2 launches there will be some modules that will have `migrate1to2` function registred

And after phase-3 upgrade, some modules will have `migrate1to2` function, and some modules will have `migrate1to2` and `migrate2to3` registered.